src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -415,6 +427,67 @@ def insertContainerRules(self):
         # Keep a list of subscriptions to tick as subscribed in the database
         subscriptionsMade = []
 
+        # Create the subscription objects and add them to the list
+        # The list takes care of the sorting internally
+        for subInfo in unsubscribedDatasets:
+            container = subInfo['path']
+            rseExpr = subInfo['site']
+            # we skip Custodial/MSS/Tape data placement
+            if subInfo['custodial'] in [1, 'y']:
+                logging.info("Bypassing custodial container rule for container: %s and RSE: %s",
+                             container, rseExpr)
+                subscriptionsMade.append(subInfo['id'])
+                continue
+
+            if not self._isContainerTierAllowed(container):
+                logging.info("Component configured to skip container rule for: %s", container)
+                subscriptionsMade.append(subInfo['id'])
+                continue
+
+            kwargs = dict(ask_approval=False, activity="Production Output",
+                          account=self.rucioAcct, grouping="ALL",
+                          comment="WMAgent automatic container rule", meta=self.metaData)
+            # top it up with component-level parameters
+            kwargs.update(self.containerDiskRuleParams)
+
+            rseExpr = self.containerDiskRuleRSEExpr
+            if self.testRSEs:
+                rseExpr = rseExpr.replace("cms_type=real", "cms_type=test")
+            logging.info("Creating container rule for %s against RSE %s", container, rseExpr)
+            try:
+                resp = self.rucio.createReplicationRule(container,
+                                                        rseExpression=rseExpr, **kwargs)
+            except Exception as exc:
```

Comment:
Looking at the code of `createReplicationRule` function, it seems that the only Exception that can be raised there and captured here is `AccessDenied` from [1]. All the rest are masked in [2], including the `DuplicateRule` few lines above. So the question is should we retry indefinitely/infinitely here expecting something to happen all by itself on the Rucio side to fix the error and give the access in question or it is worth alarming the WMCore team after few retries here that, something is wrong.   

[1]
https://github.com/dmwm/WMCore/blob/51b89389493932391080d9d4e1ad95e9356fd3bb/src/python/WMCore/Services/Rucio/Rucio.py#L505

[2]
https://github.com/dmwm/WMCore/blob/51b89389493932391080d9d4e1ad95e9356fd3bb/src/python/WMCore/Services/Rucio/Rucio.py#L533

Reply:
This code is now only dealing with Disk data placement, through a very generic rse expression. So if we hit an `AccessDenied`, it should happen only once in life, until the CMS Rucio team fixes it. Given that the impact would be very clear too - not in the form of a WMAgent alert thoguh - I decided not to deal with that specific exception in this code (it remains in the T0 logic though).

That's a good question though, and **ideally**, we should eventually retry actions a few times (over a few different cycles) until a hard failure is triggered (possibly triggering a notification/alert too). There are many places that this behaviour would be very helpful.